confgen/api_internal.go

```diff
@@ -181,7 +181,7 @@ func computeInitial(envMap EnvMap, confFilePath string, confOverrides []string,
 
 	// Fetch pertinent data from Initial Config
 
-	_, localVolumeGroupMap, localVolumeMap, _, _, err = fetchVolumeInfo(initialConfMap)
+	_, localSMBVolumeGroupMap, localVolumeMap, _, _, err = fetchVolumeInfo(initialConfMap)
```

Comment:
Is it the case that:
1) A VolumeGroup may only contain SMB or only contain NFS or only contain non-SMB/non-NFS Volumes?
2) Why must we assume fetchVolumeInfo() returns only an SMB Volume Group Map - what about NFS/non-SMB/NFS?

Reply:
Yes - I have the same questions

It seems like this function should be reorganized to take advantage of **IsVolumeGroupSharedSMB()** and **IsVolumeGroupSharedNFS()**

Only setup SMB and NFS if you have volume groups which use them.

Reply:
Currently, if a volume group has NFS volumes and/or SMB volumes then it must have a Virtual IP.

The converse is not true.  For example, it is OK for an empty volume group to have a Virtual IP.  Originally, I did add a check that a volume group without NFS or SMB volumes must not have an VIP, but I took that out when I ran into a problem with an empty VG.  Since VG will be provisioned separately from volumes, it is always possible to have such an empty VG for a short period of time.  Or if a volume gets deleted i don't want config pushes to suddenly blow up.

the answer to the second question is "expediency".  the code doesn't need to know the list of volume groups with NFS volumes, or the list of volume groups without NFS or SMB volumes because it doesn't have to do anything for them, so why bother returning them?

Bob raises an interesting point!

localSMBVolumeGroupMap is used here:
```
        // Create per VG smb.conf files                                                                                  
        err = createSMBConf(initialDirPath, localSMBVolumeGroupMap)
        if nil != err {
                // TODO - logging                                                                                        
                return
        }
```

i think i felt `createSMBConf()` should be passed only the list of SMB volume containing volume groups.  but I could, instead, have modified the for loop inside to call `IsVolumeGroupSharedSMB()`:

```
        for _, volumeGroup = range localVolumeGroupMap {
                vipDirPath = initialDirPath + "/" + vipsDirName + "/" + volumeGroup.VirtualIPAddr.String()

                err = os.Mkdir(vipDirPath, confDirPerm)
                if nil != err {
                        return
                }

                fileName := vipDirPath + "/smb-VG-" + volumeGroup.VolumeGroupName + ".conf"
                f, openErr := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, smbConfPerm)
                if nil != openErr {
                        err = openErr
                        return
                }

                err = globalTplate.Execute(f, volumeGroup)
                if nil != err {
                        return
                }

                err = sharesTplate.Execute(f, volumeGroup)
                if nil != err {
                        return
                }
        }
```

(the original problem was the `mkdir` was failing when `volumeGroup.VirtualIPAddr.String()` returned "nil".  the second attempt to create the directory "nil" failed with EEXIST.)

so i could move the `IsVolumeGroupSharedSMB()` out of `fetchVolumeInfo()` and into `createSMBConf()`.  would that make you happier?

Reply:
Generally - I think you should just make a call as to what the correct thing to do.

As a reviewer, I try to look at things from a different angle to see I can find bugs are simplify the code.

I could be entirely wrong - I just try to raise probing questions.

I always appreciate when someone looks at my code and says - “You know if you do this you may be able to simplify the code and fix this potential bug.”

Long story short - "Do the right thing!".

Reply:
I get our point about not needing to know the NFS Volumes in this case... since, as we've discussed, the /etc/exports has to have *ALL* served NFS Volumes in it whereas this function is dealing specifically with what goes in this VIPs/VG's smb.conf, right?

Reply:
Ed, you correct.

I rewrote the code so that `FetchVolumeInfo()` returns all volume groups and the `CreateSMBConf()` function then calls `IsVolumeGroupSharedSMB()`, but that had the undesirable consequence of requiring that I also pass `InitialConfMap` (the configuration map) when it was required before which meant the `templates.go` file where CreateSMBConf() lives then needed to import `github.com/swiftstack/ProxyFS/conf` whereas it didn't before.  I felt that was worse then the change making `FetchVolInfo()` less general.  I'll post the two diffs in Slack so you can compare them.

Perhaps the "right" way to fix this is to change the `VolumeGroup` struct to include booleans for NFS and SMB sharing, and then fill those in.  But that has its downsides as well.